diff --git a/build/gulpfile.vscode.ts b/build/gulpfile.vscode.ts
index 53987f88..7c7789b7 100644
--- a/build/gulpfile.vscode.ts
+++ b/build/gulpfile.vscode.ts
@@ -27,3 +27,3 @@ import { createAsar } from './lib/asar.ts';
 import minimist from 'minimist';
-import { compileBuildWithoutManglingTask, compileBuildWithManglingTask } from './gulpfile.compile.ts';
+import { compileBuildWithoutManglingTask } from './gulpfile.compile.ts';
 import { compileNonNativeExtensionsBuildTask, compileNativeExtensionsBuildTask, compileAllExtensionsBuildTask, compileExtensionMediaBuildTask, cleanExtensionsBuildTask, compileCopilotExtensionBuildTask } from './gulpfile.extensions.ts';
@@ -667,3 +667,26 @@ BUILD_TARGETS.forEach(buildTarget => {
 			);
-			vscodeTask = task.define(`vscode${dashed(platform)}${dashed(arch)}${dashed(minified)}`, task.series(
+
+			const esbuildServerTask = task.define(
+				`esbuild-vscode-reh${dashed(platform)}${dashed(arch)}${dashed(minified)}`,
+				() => runEsbuildBundle(
+					`out-vscode-reh${dashed(minified)}`,
+					!!minified,
+					true,
+					'server',
+					minified && useCdnSourceMapsForPackagingTasks ? `${sourceMappingURLBase}/core` : undefined
+				)
+			);
+
+			const esbuildServerWebTask = task.define(
+				`esbuild-vscode-reh-web${dashed(platform)}${dashed(arch)}${dashed(minified)}`,
+				() => runEsbuildBundle(
+					`out-vscode-reh-web${dashed(minified)}`,
+					!!minified,
+					true,
+					'server-web',
+					minified && useCdnSourceMapsForPackagingTasks ? `${sourceMappingURLBase}/core` : undefined
+				)
+			);
+
+			const prepackTask = task.define(`vscode${dashed(minified)}-prepack`, task.series(
 				copyCodiconsTask,
@@ -673,9 +696,24 @@ BUILD_TARGETS.forEach(buildTarget => {
 				compileExtensionMediaBuildTask,
-				writeISODate('out-build'),
 				esbuildBundleTask,
+			));
+			task.task(prepackTask);
+
+			const packingTask = task.define(`vscode${dashed(platform)}${dashed(arch)}${dashed(minified)}-packing`, task.series(
+				writeISODate('out-build'),
 				vscodeTaskCI
 			));
-		} else {
+
+			task.task(packingTask);
+
+			// Remote Extension Host related
+			task.task(task.define(`vscode-reh${dashed(platform)}${dashed(arch)}${dashed(minified)}-prepare`, task.series(esbuildServerTask)));
+			task.task(task.define(`vscode-reh-web${dashed(platform)}${dashed(arch)}${dashed(minified)}-prepare`, task.series(esbuildServerWebTask)));
+
 			vscodeTask = task.define(`vscode${dashed(platform)}${dashed(arch)}${dashed(minified)}`, task.series(
-				minified ? compileBuildWithManglingTask : compileBuildWithoutManglingTask,
+				prepackTask,
+				packingTask,
+			));
+		} else {
+			const prepackTask = task.define(`vscode${dashed(minified)}-prepack`, task.series(
+				compileBuildWithoutManglingTask,
 				cleanExtensionsBuildTask,
@@ -685,4 +723,15 @@ BUILD_TARGETS.forEach(buildTarget => {
 				minified ? minifyVSCodeTask : bundleVSCodeTask,
+			));
+			task.task(prepackTask);
+
+			const packingTask = task.define(`vscode${dashed(platform)}${dashed(arch)}${dashed(minified)}-packing`, task.series(
 				vscodeTaskCI
 			));
+
+			task.task(packingTask);
+
+			vscodeTask = task.define(`vscode${dashed(platform)}${dashed(arch)}${dashed(minified)}`, task.series(
+				prepackTask,
+				packingTask,
+			));
 		}
